### PR TITLE
update kubelet.conf after rotate certificate

### PR DESCRIPTION
**Pull request by weiwei04**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Kubeadm init creates kubelet.conf with client-certificate-data, client-key-data
as client certificate to connect to kube-apiserver, after the control plane is up & running
kubelet will rotate a new cert & key and store it in certDir/kubelet-client-(DATE|current).pem
but the kubelet.conf was never changed, so the new rotated kubelet-client-(DATE|current).pem
was never used.

I'd like to add a callback funcation to certificate_manager.go after rotate certificates also update kubelet.conf to point to new created kubelet-client-(DATE|current).pem files.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
      update kubelet.conf to point to new created kubelet-client-current.pem file
```

### CODE REVIEW

**Comment by weiwei04** [comment]
/assign @smarterclayton

Please take a look, if this code change is ok, I will add some unittest :)

**Comment by mattjmcnaughton** [comment]
@weiwei04 thanks for your work here :)

If the new certificate is named `kubelet-client-current`, I'm not sure I'm understanding the problem. If we specify `kubelet-client-current...` in the file, shouldn't that always be correct?

Can you share a little more about how certificate rotation typically happens? Does the kubelet always rotate the certificate or is this happening via a separate process?

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -853,6 +854,31 @@ func buildClientCertificateManager(certConfig, clientConfig *restclient.Config,
...
+			return err
+		}
+		ctx := config.Contexts[config.CurrentContext]
+		if ctx == nil {
+			return fmt.Errorf("invalide kubelet.conf context %s not exist", config.CurrentContext)
+		}
+		auth := config.AuthInfos[ctx.AuthInfo]
+		if auth != nil && auth.ClientCertificate == pemFilename && auth.ClientKey == pemFilename {
+			return nil
+		} else {
```
Comment (draveness) [inline cmd/kubelet/app/server.go]: This else can be removed, and we can reduce the indentation for the else body.

**Review by draveness — CHANGES_REQUESTED** [review]
/ok-to-test

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager.go [hunk 2]
```diff
@@ -160,7 +165,7 @@ type manager struct {
 	usages          []certificates.KeyUsage
 	forceRotation   bool
 
-	certStore Store
+	certStore FileStore
```
Comment (draveness) [inline staging/src/k8s.io/client-go/util/certificate/certificate_manager.go]: What is the reason to replace it with `FileStore`?

**Comment by weiwei04** [comment]
Kubelet certificate auto rotate will always name kubelet-client certificate as kubelet-client-current.pem and link it to a kubelet-client-DATE.pem.

for normal kubelet node https://github.com/kubernetes/kubernetes/pull/62152 fixed the certificate naming problem

this pr is to solve master kubelet node, which kubelet.conf is created by kubeadm and certificate is stored in certificate-client-data & certificate-client-key.

@mattjmcnaughton

**Comment by weiwei04** [comment]
resolved comments & add ut, please take a look @draveness

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager.go [hunk 3]
```diff
@@ -91,7 +94,7 @@ type Config struct {
 	// CertificateStore is a persistent store where the current cert/key is
 	// kept and future cert/key pairs will be persisted after they are
 	// generated.
-	CertificateStore Store
+	CertificateStore FileStore
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/util/certificate/certificate_manager.go]: why switch to FileStore?

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go [hunk 4]
```diff
@@ -1051,6 +1051,10 @@ func (s *fakeStore) Current() (*tls.Certificate, error) {
 	return s.cert, nil
 }
 
+func (s *fakeStore) CurrentPath() string {
+	return "kubelet-client-current.pem"
+}
+
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go]: could you please add a unit test case?

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager.go (hunk 3 above)
Comment (weiwei04) [reply staging/src/k8s.io/client-go/util/certificate/certificate_manager.go]: in https://github.com/kubernetes/kubernetes/pull/80169/files#diff-75a52f1ad830d3c146427372bebebf91R433 need use CurrentPath() to get "kubelet-client-current.pem" path, and this funcation is in FIleStore interface.

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go (hunk 4 above)
Comment (weiwei04) [reply staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go]: add a unittest for updateFn in certificate_manager_test.go

**Comment by sjenning** [comment]
NAK on this.  The kubeconfig should use the `/var/lib/kubelet/pki/kubelet-client-current.pem` symlink which is adjusted on each rotation to point at the latest client cert.

**Comment by mattjmcnaughton** [comment]
> NAK on this. The kubeconfig should use the `/var/lib/kubelet/pki/kubelet-client-current.pem` symlink which is adjusted on each rotation to point at the latest client cert.

@sjenning embarrassing lack of internet lingo knowledge on my part... what does NAK mean?

**Comment by sjenning** [comment]
My bad. NAK -> negative acknowledgement, aka -1, aka I'm not in favor of this change.
Just point your kubeconfig file at `/var/lib/kubelet/pki/kubelet-client-current.pem` and this is not needed.

**Comment by mattjmcnaughton** [comment]
On Mon, Jul 22, 2019 at 07:27:54AM -0700, Seth Jennings wrote:
> My bad. NAK -> negative acknowledgement, aka -1, aka I'm not in favor of this change.
> Just point your kubeconfig file at `/var/lib/kubelet/pki/kubelet-client-current.pem` and this is not needed.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/kubernetes/kubernetes/pull/80169#issuecomment-513814791

Haha all good - ok gotcha, yup we are on the same page :)

**Pull request closed without merging**